pallets/elections/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+        ensure_root,
+        pallet_prelude::{BlockNumberFor, OriginFor},
+    };
+    use primitives::{
+        SessionIndex, DEFAULT_MILLISECS_PER_BLOCK, DEFAULT_SESSIONS_PER_ERA, DEFAULT_SESSION_PERIOD,
+    };
+    use sp_std::{collections::btree_map::BTreeMap, prelude::Vec};
+
+    #[pallet::storage]
+    #[pallet::getter(fn members)]
+    pub type Members<T: Config> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        type DataProvider: ElectionDataProvider<Self::AccountId, Self::BlockNumber>;
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        ChangeMembers(Vec<T::AccountId>),
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight((T::BlockWeights::get().max_block, DispatchClass::Operational))]
+        pub fn change_members(origin: OriginFor<T>, members: Vec<T::AccountId>) -> DispatchResult {
+            ensure_root(origin)?;
+            Members::<T>::put(members.clone());
+            Self::deposit_event(Event::ChangeMembers(members));
+
+            Ok(())
+        }
+    }
+
+    #[pallet::type_value]
+    pub(super) fn DefaultForSessionsPerEra() -> u32 {
+        DEFAULT_SESSIONS_PER_ERA
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn sessions_per_era)]
+    pub(super) type SessionsPerEra<T: Config> =
+        StorageValue<_, SessionIndex, ValueQuery, DefaultForSessionsPerEra>;
+
+    #[pallet::type_value]
+    pub(super) fn DefaultForSessionPeriod() -> u32 {
+        DEFAULT_SESSION_PERIOD
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn session_period)]
+    pub(super) type SessionPeriod<T: Config> =
+        StorageValue<_, u32, ValueQuery, DefaultForSessionPeriod>;
+
+    #[pallet::type_value]
+    pub(super) fn DefaultForMillisecsPerBlock() -> u64 {
+        DEFAULT_MILLISECS_PER_BLOCK
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn millisecs_per_block)]
+    pub(super) type MillisecsPerBlock<T: Config> =
+        StorageValue<_, u64, ValueQuery, DefaultForMillisecsPerBlock>;
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig<T: Config> {
+        pub members: Vec<T::AccountId>,
+        pub millisecs_per_block: u64,
+        pub session_period: u32,
+        pub sessions_per_era: SessionIndex,
+    }
+
+    #[cfg(feature = "std")]
+    impl<T: Config> Default for GenesisConfig<T> {
+        fn default() -> Self {
+            Self {
+                members: Vec::new(),
+                millisecs_per_block: DEFAULT_MILLISECS_PER_BLOCK,
+                session_period: DEFAULT_SESSION_PERIOD,
+                sessions_per_era: DEFAULT_SESSIONS_PER_ERA,
+            }
+        }
+    }
```

Comment:
This pallet has some genesis config (where storage items are initialized at genesis). We add it to the runtime **not at genesis**. Are these storage items then initialized? If so, what is the mechanism here that does it?

Reply:
The pallets starts from the next block form where it got added so there is no initialization. Similarly is for pallet staking and it starts in mode `NotForcing` for new era and it's not possible for the pallet to start a new one. So we first have to set members via `change_members` and then force new era in pallet staking.

Reply:
I don't quite understand. If it is not initialized then how will the `AlephSessionApi` work? If they were not initialized right after the update then would `session_period` etc. be initialized ever?
I'm pretty sure these items will get initialized in the first block after the update, but I'm wondering how would that happen.

Reply:
Those constants are set in bin/runtime/src/lib.rs and overwritten in genesis, so here everything is ok with `AlephSessionApi`

Reply:
So in my eyes, the `session_period` call from `AlephSessionApi` reads and outputs the `session_period` storage item from pallet `Elections`. My question is why would that item exist just after the update?

Reply:
Because they are instantiated in runtime during compilation.

Reply:
Because they are instantiated in runtime during compilation.

Reply:
I think there is a misunderstanding here. How in your opinion is `Elections::session_period()` executed?

Reply:
Changed to constants.